rust/agents/scraper/src/chain_scraper/sync.rs

```diff
@@ -0,0 +1,254 @@
+                        last_valid_range_start_block = self.last_valid_range_start_block,
+                        "Found gaps in the message in range, re-indexing the same range."
+                    );
+                }
+            }
+        }
+    }
+
+    #[instrument(skip(self))]
+    async fn scrape_range(
+        &self,
+        from: u32,
+        to: u32,
+    ) -> Result<(Vec<RawMsgWithMeta>, Vec<Delivery>)> {
+        let sorted_messages = self.local.indexer.fetch_sorted_messages(from, to).await?;
+
+        let deliveries = self.deliveries(from, to).await?;
+
+        info!(
+            from,
+            to,
+            message_count = sorted_messages.len(),
+            deliveries_count = deliveries.len(),
+            "Indexed block range for chain"
+        );
+
+        let sorted_messages = sorted_messages
+            .into_iter()
+            .map(|(raw, meta)| RawMsgWithMeta { raw, meta })
+            .filter(|m| m.raw.leaf_index > self.last_leaf_index)
+            .collect::<Vec<_>>();
+
+        debug!(
+            from,
+            to,
+            message_count = sorted_messages.len(),
+            "Filtered any messages already indexed for outbox."
+        );
+
+        Ok((sorted_messages, deliveries))
+    }
+
+    /// get the deliveries for a given range from the inboxes.
+    #[instrument(skip(self))]
+    async fn deliveries(&self, from: u32, to: u32) -> Result<Vec<Delivery>> {
+        let mut delivered = vec![];
+        for (_, remote) in self.remotes.iter() {
+            debug_assert_eq!(remote.inbox.local_domain(), self.local_domain());
+            delivered.extend(
+                remote
+                    .indexer
+                    .fetch_processed_messages(from, to)
+                    .await?
+                    .into_iter()
+                    .map(|(message_hash, meta)| Delivery {
+                        inbox: remote.inbox.address(),
+                        message_hash,
+                        meta,
+                    }),
+            )
+        }
+        Ok(delivered)
+    }
+
+    #[instrument(skip(self))]
+    async fn record_data(
+        &self,
+        sorted_messages: Vec<RawMsgWithMeta>,
+        deliveries: Vec<Delivery>,
+    ) -> Result<u32> {
+        // transaction (database_id, timestamp) by transaction hash
+        let txns: HashMap<H256, TxnWithIdAndTime> = self
+            .ensure_blocks_and_txns(
+                sorted_messages
+                    .iter()
+                    .map(|r| &r.meta)
+                    .chain(deliveries.iter().map(|d| &d.meta)),
+            )
+            .await?
+            .map(|t| (t.hash, t))
+            .collect();
+
+        let max_leaf_index_of_batch = self.store_messages(&sorted_messages, &txns).await?;
+        self.stored_messages.inc_by(sorted_messages.len() as u64);
+        self.record_deliveries(&deliveries, &txns).await?;
+        self.stored_deliveries.inc_by(deliveries.len() as u64);
```

Comment:
Yes, we have discussed this before:
https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1159#discussion_r1006277654

Basically it is meant to a count of how many we store, not how many are stored. This is by far the easiest way to track the metric and also is more useful as we might be performing updates when we duplicate an insert. If you want an actual count of things in the database that will take some work to fix not just here but in all the agents.

Reply:
Oh, sorry reading a little closer, we might perform updates; there might be a way to get a returned value of just how many were inserted/updated; let me look into this.

Reply:
Yeah, looks like the best I can do is just get full models returned which is kinda costly since all we need is the count. This is a sea-orm limitation but fwiw doing the metric this way does still represent to a degree how much we are hitting the db with.

Reply:
Yeah happy to keep as is